src/codecs/md/index.ts

```diff
@@ -1011,26 +1022,73 @@ function encodeFigure(figure: stencila.Figure): Extension {
 }
 
 /**
- * Decode a `chunkfigure` block extension to a `Figure` with a `CodeChunk`
- * as it's first child.
+ * Decode a `chunkfigure` block extension to a `Figure` node.
  */
-function decodeCodeChunkFigure(ext: Extension): stencila.Figure {
+function decodeChunkFigure(ext: Extension): stencila.Figure {
   const { content = '' } = ext
-  const [codeAndOutputs, caption = ''] = content.split('---')
-  const chunk = decodeCodeChunk({
-    ...ext,
-    content: codeAndOutputs,
-  })
+  const [figureObject, caption = ''] = content.split(/^---$/m)
+
+  let children: stencila.Node = []
+  // Always encode `Code` Blocks to `CodeChunk` nodes
+  if (figureObject.includes('```')) {
+    children = decodeCodeChunk({
+      ...ext,
+      content: figureObject,
+    })
+  } else {
+    const decoded = decodeMarkdown(figureObject).content
+    children = decoded ? decoded[0] : []
+  }
+
   const figure = decodeFigure({
     ...ext,
     content: `dummy\n\n${caption}`,
   })
   if (figure.content !== undefined) {
-    figure.content[0] = chunk
+    figure.content[0] = children
   }
   return figure
 }
 
+/**
+ * Encode a `figure` node to a `chunkfigure` block extension
+ */
+function encodeChunkFigure(figure: stencila.Figure): UNIST.Node[] {
+  const chunkFigure: UNIST.Node[] = []
+  let content = ``
+
+  const figureContent = figure.content?.length === 1 && figure.content[0]
+
+  if (isA('CodeChunk', figureContent)) {
+    // Special handling for encoding `CodeChunks`. Otherwise it leads to double nesting of block extensions
+    content += encodeCodeChunk(figureContent).content ?? ''
+  } else {
+    content += encodeMarkdown(figureContent)
+  }
+
+  content += '\n---\n'
+
+  content +=
+    typeof figure.caption === 'string'
+      ? figure.caption
+      : Array.isArray(figure.caption)
+      ? figure.caption.map(encodeMarkdown).join('\n\n')
+      : ''
+
+  chunkFigure.push({
+    type: 'block-extension',
+    name: 'chunkfigure',
+    argument: figure.label,
+    content,
+  })
+
+  if (figure.id) {
+    chunkFigure.push(encodeString(`{#${figure.id}}`))
+  }
```

Comment:
This is why the space is appearing above between the block extension and the id. Instead, the id should be part of the block extension's `properties` i.e.

```js
  chunkFigure.push({
    type: 'block-extension',
    name: 'chunkfigure',
    argument: figure.label,
    properties: {id},
    content,
  })
```